File: android/lca/src/main/java/com/dreampany/lca/data/model/Quote.java
/*
package com.dreampany.lca.data.model;

import android.os.Parcel;

import androidx.annotation.NonNull;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;

import com.dreampany.frame.data.model.Base;
import com.dreampany.lca.data.enums.Currency;
import com.dreampany.lca.misc.Constants;
import com.google.firebase.database.IgnoreExtraProperties;
import com.google.firebase.database.PropertyName;

import org.apache.commons.lang3.builder.ReflectionToStringBuilder;

*/
/**
 * Created by Roman on 1/14/2019
 * Copyright (c) 2019 Dreampany. All rights reserved.
 * dreampanymail@gmail.com
 * Last modified $file.lastModified
 *//*


@Entity(indices = {@Index(value = {Constants.Quote.ID, Constants.Quote.CURRENCY}, unique = true)},
        primaryKeys = {Constants.Quote.ID, Constants.Quote.CURRENCY})
@IgnoreExtraProperties
public class Quote extends Base {

    @NonNull
    private Currency currency;
    private double price;
    @PropertyName(Constants.Quote.DAY_VOLUME)
    private double dayVolume;
    @PropertyName(Constants.Quote.MARKET_CAP)
    private double marketCap;
    @PropertyName(Constants.Quote.HOUR_CHANGE)
    private double hourChange;
    @PropertyName(Constants.Quote.DAY_CHANGE)
    private double dayChange;
    @PropertyName(Constants.Quote.WEEK_CHANGE)
    private double weekChange;
    @PropertyName(Constants.Quote.LAST_UPDATED)
    private long lastUpdated;

    @Ignore
    public Quote() {

    }

    public Quote(String id) {
        this.id = id;
    }

    @Ignore
    private Quote(Parcel in) {
        super(in);
        currency = in.readParcelable(Currency.class.getClassLoader());
        price = in.readDouble();
        dayVolume = in.readDouble();
        marketCap = in.readDouble();
        hourChange = in.readDouble();
        dayChange = in.readDouble();
        weekChange = in.readDouble();
        lastUpdated = in.readLong();
    }

    @Override
    public void writeToParcel(@NonNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeParcelable(currency, flags);
        dest.writeDouble(price);
        dest.writeDouble(dayVolume);
        dest.writeDouble(marketCap);
        dest.writeDouble(hourChange);
        dest.writeDouble(dayChange);
        dest.writeDouble(weekChange);
        dest.writeLong(lastUpdated);
    }

    public static final Creator<Quote> CREATOR = new Creator<Quote>() {
        @Override
        public Quote createFromParcel(Parcel in) {
            return new Quote(in);
        }

        @Override
        public Quote[] newArray(int size) {
            return new Quote[size];
        }
    };

    @NonNull
    @Override
    public String toString() {
        return ReflectionToStringBuilder.toString(this);
    }

    public void setCurrency(@NonNull Currency currency) {
        this.currency = currency;
    }

    @NonNull
    public Currency getCurrency() {
        return currency;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public double getPrice() {
        return price;
    }

    @PropertyName(Constants.Quote.DAY_VOLUME)
    public void setDayVolume(double dayVolume) {
        this.dayVolume = dayVolume;
    }

    @PropertyName(Constants.Quote.DAY_VOLUME)
    public double getDayVolume() {
        return dayVolume;
    }

    @PropertyName(Constants.Quote.MARKET_CAP)
    public void setMarketCap(double marketCap) {
        this.marketCap = marketCap;
    }

    @PropertyName(Constants.Quote.MARKET_CAP)
    public double getMarketCap() {
        return marketCap;
    }

    @PropertyName(Constants.Quote.HOUR_CHANGE)
    public void setHourChange(double hourChange) {
        this.hourChange = hourChange;
    }

    @PropertyName(Constants.Quote.HOUR_CHANGE)
    public double getHourChange() {
        return hourChange;
    }

    @PropertyName(Constants.Quote.DAY_CHANGE)
    public void setDayChange(double dayChange) {
        this.dayChange = dayChange;
    }

    @PropertyName(Constants.Quote.DAY_CHANGE)
    public double getDayChange() {
        return dayChange;
    }

    @PropertyName(Constants.Quote.WEEK_CHANGE)
    public void setWeekChange(double weekChange) {
        this.weekChange = weekChange;
    }

    @PropertyName(Constants.Quote.WEEK_CHANGE)
    public double getWeekChange() {
        return weekChange;
    }

    @PropertyName(Constants.Quote.LAST_UPDATED)
    public void setLastUpdated(long lastUpdated) {
        this.lastUpdated = lastUpdated;
    }

    @PropertyName(Constants.Quote.LAST_UPDATED)
    public long getLastUpdated() {
        return lastUpdated;
    }
}
*/

File: android/lca/src/main/kotlin/com/dreampany/lca/data/model/Quote.kt
package com.dreampany.lca.data.model

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import com.dreampany.framework.data.model.Base
import com.dreampany.framework.util.TimeUtilKt
import com.dreampany.lca.data.enums.Currency
import com.dreampany.lca.misc.Constants
import com.google.common.base.Objects
import com.google.firebase.database.IgnoreExtraProperties
import com.google.firebase.database.PropertyName
import kotlinx.android.parcel.Parcelize

/**
 * Created by Roman-372 on 8/2/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Parcelize
@IgnoreExtraProperties
@Entity(
    indices = [Index(
        value = [Constants.Quote.ID, Constants.Quote.CURRENCY],
        unique = true
    )],
    primaryKeys = [Constants.Quote.ID, Constants.Quote.CURRENCY]
)
data class Quote(
    override var time: Long = Constants.Default.LONG,
    override var id: String = Constants.Default.STRING,
    var currency: Currency = Currency.USD,
    var price: Double = Constants.Default.DOUBLE,
    private var dayVolume: Double = Constants.Default.DOUBLE,
    private var marketCap: Double = Constants.Default.DOUBLE,
    private var hourChange: Double = Constants.Default.DOUBLE,
    private var dayChange: Double = Constants.Default.DOUBLE,
    private var weekChange: Double = Constants.Default.DOUBLE,
    private var lastUpdated: Long = 0
) : Base() {

    @Ignore
    constructor() : this(TimeUtilKt.currentMillis()) {
    }

    constructor(id: String) : this(time = TimeUtilKt.currentMillis(), id = id) {}

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val item = other as Quote
        return Objects.equal(item.id, id) && Objects.equal(item.currency, currency)
    }

    override fun hashCode(): Int {
        return Objects.hashCode(id, currency)
    }

    @PropertyName(Constants.Quote.DAY_VOLUME)
    fun setDayVolume(dayVolume: Double) {
        this.dayVolume = dayVolume
    }

    @PropertyName(Constants.Quote.DAY_VOLUME)
    fun getDayVolume(): Double {
        return dayVolume
    }

    @PropertyName(Constants.Quote.MARKET_CAP)
    fun setMarketCap(marketCap: Double) {
        this.marketCap = marketCap
    }

    @PropertyName(Constants.Quote.MARKET_CAP)
    fun getMarketCap(): Double {
        return marketCap
    }

    @PropertyName(Constants.Quote.HOUR_CHANGE)
    fun setHourChange(hourChange: Double) {
        this.hourChange = hourChange
    }

    @PropertyName(Constants.Quote.HOUR_CHANGE)
    fun getHourChange(): Double {
        return hourChange
    }

    @PropertyName(Constants.Quote.DAY_CHANGE)
    fun setDayChange(dayChange: Double) {
        this.dayChange = dayChange
    }

    @PropertyName(Constants.Quote.DAY_CHANGE)
    fun getDayChange(): Double {
        return dayChange
    }

    @PropertyName(Constants.Quote.WEEK_CHANGE)
    fun setWeekChange(weekChange: Double) {
        this.weekChange = weekChange
    }

    @PropertyName(Constants.Quote.WEEK_CHANGE)
    fun getWeekChange(): Double {
        return weekChange
    }

    @PropertyName(Constants.Quote.LAST_UPDATED)
    fun setLastUpdated(lastUpdated: Long) {
        this.lastUpdated = lastUpdated
    }

    @PropertyName(Constants.Quote.LAST_UPDATED)
    fun getLastUpdated(): Long {
        return lastUpdated
    }
}